File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Conformance/PressureRunner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json.Nodes;
using Iaiso.Audit;
using Iaiso.Core;

namespace Iaiso.Conformance;

internal static class PressureRunner
{
    public static List<VectorResult> Run(string specRoot)
    {
        var path = Path.Combine(specRoot, "pressure", "vectors.json");
        var doc = JsonNode.Parse(File.ReadAllBytes(path))!.AsObject();
        var vectors = doc["vectors"]!.AsArray();
        var results = new List<VectorResult>(vectors.Count);
        foreach (var v in vectors)
        {
            results.Add(RunOne(v!.AsObject()));
        }
        return results;
    }

    private static VectorResult RunOne(JsonObject v)
    {
        string name = v["name"]!.GetValue<string>();

        var cb = PressureConfig.CreateBuilder();
        if (v["config"] is JsonObject c)
        {
            ApplyDouble(c, "escalation_threshold", x => cb.EscalationThreshold = x);
            ApplyDouble(c, "release_threshold", x => cb.ReleaseThreshold = x);
            ApplyDouble(c, "dissipation_per_step", x => cb.DissipationPerStep = x);
            ApplyDouble(c, "dissipation_per_second", x => cb.DissipationPerSecond = x);
            ApplyDouble(c, "token_coefficient", x => cb.TokenCoefficient = x);
            ApplyDouble(c, "tool_coefficient", x => cb.ToolCoefficient = x);
            ApplyDouble(c, "depth_coefficient", x => cb.DepthCoefficient = x);
            if (c["post_release_lock"] is JsonValue prl && prl.TryGetValue<bool>(out var b))
                cb.PostReleaseLock = b;
        }
        var cfg = cb.Build();

        double[] clockSeq;
        if (v["clock"] is JsonArray clkArr)
        {
            clockSeq = new double[clkArr.Count];
            for (int i = 0; i < clkArr.Count; i++) clockSeq[i] = clkArr[i]!.GetValue<double>();
        }
        else
        {
            clockSeq = new[] { 0.0 };
        }
        var clock = new ScriptedClock(clockSeq);

        string? expectErr = v["expect_config_error"]?.GetValue<string>();
        PressureEngine engine;
        try
        {
            engine = new PressureEngine(cfg, new EngineOptions
            {
                ExecutionId = "vec-" + name,
                AuditSink = NullSink.Instance,
                Clock = clock,
                TimestampClock = clock,
            });
        }
        catch (ConfigException e)
        {
            if (expectErr is null)
                return VectorResult.Fail("pressure", name,
                    "engine construction failed: " + e.Message);
            if (!e.Message.Contains(expectErr))
                return VectorResult.Fail("pressure", name,
                    $"expected error containing '{expectErr}', got: {e.Message}");
            return VectorResult.Pass("pressure", name);
        }
        if (expectErr is not null)
            return VectorResult.Fail("pressure", name,
                $"expected config error containing '{expectErr}', got Ok");

        if (v["expected_initial"] is JsonObject init)
        {
            var snap = engine.Snapshot();
            if (Math.Abs(snap.Pressure - init["pressure"]!.GetValue<double>()) > ConformanceRunner.Tolerance)
                return VectorResult.Fail("pressure", name,
                    $"initial pressure: got {snap.Pressure}, want {init["pressure"]!.GetValue<double>()}");
            if (snap.Step != init["step"]!.GetValue<long>())
                return VectorResult.Fail("pressure", name,
                    $"initial step: got {snap.Step}, want {init["step"]!.GetValue<long>()}");
            if (snap.Lifecycle.Wire() != init["lifecycle"]!.GetValue<string>())
                return VectorResult.Fail("pressure", name,
                    $"initial lifecycle: got {snap.Lifecycle.Wire()}, want {init["lifecycle"]!.GetValue<string>()}");
            if (Math.Abs(snap.LastStepAt - init["last_step_at"]!.GetValue<double>()) > ConformanceRunner.Tolerance)
                return VectorResult.Fail("pressure", name,
                    $"initial last_step_at: got {snap.LastStepAt}, want {init["last_step_at"]!.GetValue<double>()}");
        }

        var steps = v["steps"] as JsonArray ?? new JsonArray();
        var expSteps = v["expected_steps"] as JsonArray ?? new JsonArray();
        for (int i = 0; i < steps.Count; i++)
        {
            var step = steps[i]!.AsObject();
            StepOutcome outcome;
            if (step["reset"] is JsonValue rv && rv.TryGetValue<bool>(out var resetB) && resetB)
            {
                engine.Reset();
                outcome = StepOutcome.Ok;
            }
            else
            {
                var si = new StepInput();
                if (step["tokens"] is JsonValue tv) si.Tokens = tv.GetValue<long>();
                if (step["tool_calls"] is JsonValue tcv) si.ToolCalls = tcv.GetValue<long>();
                if (step["depth"] is JsonValue dv) si.Depth = dv.GetValue<long>();
                if (step["tag"] is JsonValue tg && tg.TryGetValue<string>(out var tag)) si.Tag = tag;
                outcome = engine.Step(si);
            }
            if (i >= expSteps.Count)
                return VectorResult.Fail("pressure", name,
                    $"step {i}: no expected entry");
            var exp = expSteps[i]!.AsObject();
            if (outcome.Wire() != exp["outcome"]!.GetValue<string>())
                return VectorResult.Fail("pressure", name,
                    $"step {i}: outcome got {outcome.Wire()}, want {exp["outcome"]!.GetValue<string>()}");
            var snap = engine.Snapshot();
            if (Math.Abs(snap.Pressure - exp["pressure"]!.GetValue<double>()) > ConformanceRunner.Tolerance)
                return VectorResult.Fail("pressure", name,
                    $"step {i}: pressure got {snap.Pressure}, want {exp["pressure"]!.GetValue<double>()}");
            if (snap.Step != exp["step"]!.GetValue<long>())
                return VectorResult.Fail("pressure", name,
                    $"step {i}: step got {snap.Step}, want {exp["step"]!.GetValue<long>()}");
            if (snap.Lifecycle.Wire() != exp["lifecycle"]!.GetValue<string>())
                return VectorResult.Fail("pressure", name,
                    $"step {i}: lifecycle got {snap.Lifecycle.Wire()}, want {exp["lifecycle"]!.GetValue<string>()}");
        }
        return VectorResult.Pass("pressure", name);
    }

    private static void ApplyDouble(JsonObject obj, string key, Action<double> setter)
    {
        if (obj[key] is JsonValue v && v.TryGetValue<double>(out var d)) setter(d);
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/conformance/pressure_runner.rb
# frozen_string_literal: true

require "json"
require_relative "result"
require_relative "../audit/null_sink"
require_relative "../core/pressure_config"
require_relative "../core/pressure_engine"
require_relative "../core/clock"

module IAIso
  module Conformance
    module PressureRunner
      TOLERANCE = 1e-9

      module_function

      def run(spec_root)
        path = File.join(spec_root, "pressure", "vectors.json")
        doc = ::JSON.parse(File.read(path))
        doc["vectors"].map { |v| run_one(v) }
      end

      def run_one(v)
        name = v["name"] || "?"

        # Build PressureConfig with overrides from vector.
        kw = {}
        if v["config"].is_a?(Hash)
          c = v["config"]
          %w[escalation_threshold release_threshold dissipation_per_step
             dissipation_per_second token_coefficient tool_coefficient
             depth_coefficient].each do |f|
            kw[f.to_sym] = c[f].to_f if c.key?(f) && c[f].is_a?(Numeric)
          end
          kw[:post_release_lock] = c["post_release_lock"] if [true, false].include?(c["post_release_lock"])
        end

        cfg = IAIso::Core::PressureConfig.new(**kw)
        clock_seq = (v["clock"].is_a?(Array) ? v["clock"] : [0.0]).map(&:to_f)
        clk = IAIso::Core::ScriptedClock.new(clock_seq)

        expect_err = v["expect_config_error"]
        engine = nil
        begin
          engine = IAIso::Core::PressureEngine.new(
            config: cfg,
            options: IAIso::Core::EngineOptions.new(
              execution_id: "vec-#{name}",
              audit_sink: IAIso::Audit::NullSink.instance,
              clock: clk, timestamp_clock: clk,
            ),
          )
        rescue IAIso::Core::ConfigError => e
          if expect_err.nil?
            return VectorResult.fail("pressure", name, "engine init failed: #{e.message}")
          end
          unless e.message.include?(expect_err)
            return VectorResult.fail("pressure", name, "expected error containing '#{expect_err}', got: #{e.message}")
          end
          return VectorResult.pass("pressure", name)
        end
        if expect_err
          return VectorResult.fail("pressure", name, "expected config error containing '#{expect_err}', got Ok")
        end

        # initial state check
        if v["expected_initial"].is_a?(Hash)
          init = v["expected_initial"]
          snap = engine.snapshot
          if (snap.pressure - init["pressure"].to_f).abs > TOLERANCE
            return VectorResult.fail("pressure", name, "initial pressure: got #{snap.pressure}, want #{init["pressure"]}")
          end
          if snap.step != init["step"].to_i
            return VectorResult.fail("pressure", name, "initial step: got #{snap.step}, want #{init["step"]}")
          end
          if snap.lifecycle != init["lifecycle"]
            return VectorResult.fail("pressure", name, "initial lifecycle: got #{snap.lifecycle}, want #{init["lifecycle"]}")
          end
          if (snap.last_step_at - init["last_step_at"].to_f).abs > TOLERANCE
            return VectorResult.fail("pressure", name, "initial last_step_at: got #{snap.last_step_at}, want #{init["last_step_at"]}")
          end
        end

        # steps
        steps = v["steps"] || []
        exp_steps = v["expected_steps"] || []
        steps.each_with_index do |step, i|
          if step["reset"]
            engine.reset
            outcome = "ok"
          else
            kw = {}
            kw[:tokens] = step["tokens"].to_i if step.key?("tokens")
            kw[:tool_calls] = step["tool_calls"].to_i if step.key?("tool_calls")
            kw[:depth] = step["depth"].to_i if step.key?("depth")
            kw[:tag] = step["tag"] if step.key?("tag")
            outcome = engine.step(IAIso::Core::StepInput.build(**kw))
          end
          exp = exp_steps[i]
          unless exp
            return VectorResult.fail("pressure", name, "step #{i}: no expected entry")
          end
          if outcome != exp["outcome"]
            return VectorResult.fail("pressure", name, "step #{i}: outcome got #{outcome}, want #{exp["outcome"]}")
          end
          snap = engine.snapshot
          if (snap.pressure - exp["pressure"].to_f).abs > TOLERANCE
            return VectorResult.fail("pressure", name, "step #{i}: pressure got #{snap.pressure}, want #{exp["pressure"]}")
          end
          if snap.step != exp["step"].to_i
            return VectorResult.fail("pressure", name, "step #{i}: step got #{snap.step}, want #{exp["step"]}")
          end
          if snap.lifecycle != exp["lifecycle"]
            return VectorResult.fail("pressure", name, "step #{i}: lifecycle got #{snap.lifecycle}, want #{exp["lifecycle"]}")
          end
        end
        VectorResult.pass("pressure", name)
      end
    end
  end
end
